### PR TITLE
MemoryCache: Size calculation with expired Entries

### LINKED ISSUE #36037: MemoryCache Size calculation fails when entry to add is already expired
When using MemoryCache with SizeLimit, adding an entry with past expiration will cause the Cache Size to be incremented during a call to `UpdateCacheSizeExceedsCapacity` but it is never decremented (l.207ff.).

Repro:
```
using System;
using System.Diagnostics.CodeAnalysis;
using System.Reflection;
using System.Threading;
using Microsoft.Extensions.Caching.Memory;

namespace CacheOverflow
{
	internal class Program
	{
		private readonly MemoryCache memoryCache;

		private Program()
		{
			this.memoryCache = new MemoryCache( new MemoryCacheOptions { SizeLimit = 5 } );
		}

		private static void Main( string[] args )
		{
			var p = new Program();
			p.Start();
			Console.ReadLine();
		}

		private void Start()
		{
			for ( var i = 0; i < 10; i++ )
			{
				var entryOptions = new MemoryCacheEntryOptions
				{
					AbsoluteExpiration = DateTimeOffset.UtcNow.AddMinutes( -1 ),
					Size = 1
				};
				entryOptions.RegisterPostEvictionCallback( this.EvictionCallback );
				this.memoryCache.Set( i, i, entryOptions );
				Console.WriteLine( $"Cache SET: '{i}'. Cache Size: {MemoryCacheDelegates.GetCacheSize(this.memoryCache)}. Object count: '{this.memoryCache.Count}'");
				Thread.Sleep( 100 );
			}
		}

		private void EvictionCallback( object key,
			object value,
			EvictionReason reason,
			object state )
		{
			Console.WriteLine( $"Evicted key '{key}' for reason '{reason}'. Cache Size: {MemoryCacheDelegates.GetCacheSize(this.memoryCache)}. Object count: '{this.memoryCache.Count}'");
		}
	}

	public static class MemoryCacheDelegates
	{
		[NotNull]
		public static readonly Func<MemoryCache, long> GetCacheSize;

		static MemoryCacheDelegates()
		{
			var getter = typeof( MemoryCache ).GetProperty( "Size", BindingFlags.NonPublic | BindingFlags.Instance )?.GetMethod;
			GetCacheSize = getter?.CreateDelegate( typeof( Func<MemoryCache, long> ) ) as Func<MemoryCache, long> ?? (cache => -1L);
		}
	}
}

```

### PR DESCRIPTION
**Pull request by Bernd-Broesel**
Trying to add an already expired entry to the cache no longer
permanently increases the cache size

Fix #36037

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189) before we can accept your contribution.<br/><br/>:x: hiersekornc [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=39189)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=39189) it.</sub>

File: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs [hunk 1]
```diff
@@ -341,6 +341,23 @@ public async Task ExpiringEntryDecreasesCacheSize()
             Assert.Equal(0, cache.Size);
         }
 
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/36037")]
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:345]: Please remove `ActiveIssue`.  This attribute is used to disable a test against an issue when the test is failing.   You don't need that here.
```suggestion
```

**Review by ericstj — COMMENTED** [review]
Looks good, just one small suggestion.

File: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs (hunk 1 above)
Comment (hiersekornc) [reply src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs]: Hi Eric, thanks for your review. The attribute has been removed. Additionally I have added a test for an expired token.

File: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs [hunk 2]
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
...
+            cache.Set("key", "value", entryOptions);
+            
+            Assert.Null(cache.Get("key"));
+            Assert.Equal(0, cache.Size);
+        }
+
+        [Fact]
+        public void TryingToAddEntryWithExpiredTokenDoesNotIncreaseCacheSize()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions {SizeLimit = 10});
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:364]: nit: formatting
```suggestion
            var cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = 10 });
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs [hunk 3]
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
             Assert.Equal(0, cache.Size);
         }
 
+        [Fact]
+        public void TryingToAddExpiredEntryDoesNotIncreaseCacheSize()
+        {
+            var testClock = new TestClock();
+            var cache = new MemoryCache(new MemoryCacheOptions {Clock = testClock, SizeLimit = 10});
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:348]: nit: formatting
```suggestion
            var cache = new MemoryCache(new MemoryCacheOptions { Clock = testClock, SizeLimit = 10 });
```

**Review by ericstj — APPROVED** [review]
LGTM with minor code formatting suggestions.  Please change from draft->official PR.   I'd also like a review from extensions-Caching owners @maryamariyan @eerhardt

File: src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs [hunk 4]
```diff
@@ -341,6 +341,39 @@ public async Task ExpiringEntryDecreasesCacheSize()
...
+        }
+
+        [Fact]
+        public void TryingToAddEntryWithExpiredTokenDoesNotIncreaseCacheSize()
+        {
+            var cache = new MemoryCache(new MemoryCacheOptions {SizeLimit = 10});
+            var testExpirationToken = new TestExpirationToken { HasChanged = true };
+            var entryOptions = new MemoryCacheEntryOptions
+            {
+                Size = 5, ExpirationTokens = { testExpirationToken }
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Caching.Memory/tests/CapacityTests.cs:368]: nit: formatting
```suggestion
                Size = 5,
                ExpirationTokens = { testExpirationToken }
```

**Comment by hiersekornc** [comment]
Formatting suggestions have been applied.

**Review by maryamariyan — APPROVED** [review]

**Comment by maryamariyan** [comment]
@hiersekornc thank you for your contribution!

**Pull request merged**